oak/server/rust/oak_runtime/src/metrics/mod.rs

```diff
@@ -15,53 +15,127 @@
-                "The gRPC request latencies in seconds."
-            )
-            .expect("Creating grpc_request_duration_seconds metric failed."),
-
-            grpc_requests_total: register_int_counter!(opts!(
-                "grpc_requests_total",
-                "Total number of gRPC requests received."
-            ))
-            .expect("Creating grpc_requests_total metric failed."),
-
-            grpc_response_size: register_histogram!(
+            registry: Registry::new(),
+        }
+    }
+
+    fn register<T: 'static + prometheus::core::Collector + Clone>(&self, metric: T) -> T {
+        self.registry.register(Box::new(metric.clone())).unwrap();
+        metric
+    }
+}
+
+fn counter(metric_name: &str, labels: &[&str], help: &str) -> IntCounterVec {
+    let opts = Opts::new(metric_name, help);
+    IntCounterVec::new(opts, labels).unwrap()
+}
+
+fn histogram(metric_name: &str, label: &[&str], help: &str) -> HistogramVec {
+    let opts = HistogramOpts::new(metric_name, help);
+    HistogramVec::new(opts, label).unwrap()
+}
+
+fn int_gauge(metric_name: &str, label: &[&str], help: &str) -> IntGaugeVec {
+    let opts = Opts::new(metric_name, help);
+    IntGaugeVec::new(opts, label).unwrap()
+}
+
+impl GrpcServerMetrics {
+    fn new(builder: &MetricsBuilder) -> Self {
+        GrpcServerMetrics {
+            grpc_server_started_total: builder.register(counter(
+                "grpc_server_started_total",
+                &["method_name"],
+                "Total number of RPCs started on the server.",
+            )),
+            grpc_server_handled_total: builder.register(counter(
+                "grpc_server_handled_total",
+                &["method_name", "status_code"],
+                "Total number of RPCs completed on the server, regardless of success or failure.",
+            )),
+            grpc_server_handled_latency_seconds: builder.register(histogram(
+                "grpc_server_handled_latency_seconds",
+                &["method_name"],
+                "Histogram of response latency of RPCs handled by the server.",
+            )),
+            grpc_response_size_bytes: builder.register(histogram(
                 "grpc_response_size_bytes",
-                "The gRPC response sizes in bytes."
-            )
-            .expect("Creating grpc_response_size_bytes metric failed."),
-
-            runtime_nodes_count: register_int_gauge!(opts!(
-                "runtime_nodes_count",
-                "Number of nodes in the runtime."
-            ))
-            .expect("Creating runtime_nodes_count metric failed."),
+                &["method_name"],
+                "Histogram of response sizes of RPCs handled by the server.",
+            )),
+        }
+    }
+}
+
+impl RuntimeMetrics {
+    fn new(builder: &MetricsBuilder) -> Self {
+        RuntimeMetrics {
+            runtime_nodes_total: builder.register(int_gauge(
+                "runtime_nodes_total",
+                &["instance"],
```

Comment:
What is this label for? AFAICT it's always set to `runtime`?

Reply:
I wanted to keep everything as vector. With vector metrics, label is needed. I have now changed the type to `IntGauge` instead of `IntGaugeVec`.